// @utils-npm/thunk-utils/src/Const/ActionTypeSuffixes.ts
/** Суффикс открытия модального окна */
export const CLEAR_RELATIONS_COLLECTIONS = '_CLEAR_RELATIONS_COLLECTIONS';

/** Суффикс указывающий на неполное обновление данных */
export const PARTIAL = '_PARTIAL';

/** Суффикс добавления элемента */
export const ADD_ITEM = '_ADD_ITEM';

/** Суффикс очищения элемента */
export const CLEAR_ITEM = '_CLEAR_ITEM';

/** Суффикс копирования элемента */
export const COPY_ITEM = '_COPY_ITEM';

/** Суффикс создания элемента */
export const CREATE_ITEM = '_CREATE_ITEM';

/** Суффикс удаления элемента */
export const DELETE_ITEM = '_DELETE_ITEM';

/** Суффикс получения элемента */
export const GET_ITEM = '_GET_ITEM';

/** Суффикс загрузки элемента */
export const LOAD_ITEM = '_LOAD_ITEM';

/** Суффикс создания элемента */
export const POST_ITEM = '_POST_ITEM';

/** Суффикс удаления элемента */
export const REMOVE_ITEM = '_REMOVE_ITEM';

/** Суффикс отправки элемента */
export const SEND_ITEM = '_SEND_ITEM';

/** Суффикс установки элемента */
export const SET_ITEM = '_SET_ITEM';

/** Суффикс обновления элемента */
export const UPDATE_ITEM = '_UPDATE_ITEM';

/** Суффикс очистки списка */
export const CLEAR_LIST = '_CLEAR_LIST';

/** Суффикс создания списка */
export const CREATE_LIST = '_CREATE_LIST';

/** Суффикс загрузки списка */
export const LOAD_LIST = '_LOAD_LIST';

/** Суффикс отправки списка */
export const SEND_LIST = '_SEND_LIST';

/** Суффикс закрытия модального окна */
export const CLOSE_MODAL = '_CLOSE_MODAL';

/** Суффикс открытия модального окна */
export const OPEN_MODAL = '_OPEN_MODAL';

/** Суффикс загрузки шаблона */
export const LOAD_PREFILL = '_LOAD_PREFILL';

/** Суффикс загрузки детальной информации */
export const LOAD_DETAILS = '_LOAD_DETAILS';
